### ISSUE TITLE
Support for  Flash-Decoding:speed up long-context inference 

### ISSUE THREAD

**Issue by shiqingzhangCSU** [NVIDIA/TensorRT-LLM#209]
Flash-Decoding can speeds up attention during inference, bringing up to 8x faster generation for very long sequences. 
and Flash-Decoding is powerful compare to flash-attention during inference time. Is there any plan for support?  Thanks！
![image](https://github.com/NVIDIA/TensorRT-LLM/assets/44834482/0571d180-2bfb-4ea3-9862-4828f86636e3)

**Comment by jdemouth-nvidia** [comment]
Hi @shiqingzhangCSU ,

An approach similar to Flash Decoding is already supported in TensorRT-LLM. It’s called multi-block mode in the Attention module. 

We have not extensively tested that feature so I recommend you to test it with care but we’ll do more work to enable it by default soon. 

As of now, we are investigating a bug with that feature as one of our partners runs into problems when enabling it. It’ll be fixed soon. 

Thanks,
Julien

**Comment by hchoi-moveworks** [comment]
Thanks @jdemouth-nvidia  for the detailed explanation!

Is `multi_block_mode` available for Llama2 model as well?
It seems like it is available for `ChatGLM` but not Llama2 model
https://github.com/NVIDIA/TensorRT-LLM/blob/11e14500f35dd781b535ba009c906f55ecfee3b5/tensorrt_llm/models/chatglm6b/model.py#L49

I guess to add Llama2 support we would need to pass in `multi_block_mode` when we are creating attention layer right?

https://github.com/NVIDIA/TensorRT-LLM/blob/11e14500f35dd781b535ba009c906f55ecfee3b5/tensorrt_llm/models/llama/model.py#L66
https://github.com/NVIDIA/TensorRT-LLM/blob/11e14500f35dd781b535ba009c906f55ecfee3b5/tensorrt_llm/layers/attention.py#L136

**Comment by byshiue** [comment]
> Thanks @jdemouth-nvidia for the detailed explanation!
> 
> Is `multi_block_mode` available for Llama2 model as well? It seems like it is available for `ChatGLM` but not Llama2 model
> 
> https://github.com/NVIDIA/TensorRT-LLM/blob/11e14500f35dd781b535ba009c906f55ecfee3b5/tensorrt_llm/models/chatglm6b/model.py#L49
> 
> I guess to add Llama2 support we would need to pass in `multi_block_mode` when we are creating attention layer right?
> 
> https://github.com/NVIDIA/TensorRT-LLM/blob/11e14500f35dd781b535ba009c906f55ecfee3b5/tensorrt_llm/models/llama/model.py#L66
> 
> 
> https://github.com/NVIDIA/TensorRT-LLM/blob/11e14500f35dd781b535ba009c906f55ecfee3b5/tensorrt_llm/layers/attention.py#L136

`multi_block_mode` option is not exposed in all models, but you should be able to enable it in any model.

**Comment by hchoi-moveworks** [comment]
Thanks @byshiue !

Will try to add support for Llama to test this feature out. 
Would we also need to change logic for converting huggingface checkpoint to Trt checkpoint? [1] or would just checking the `build.py`[2] to pass in `multi_block_mode` parameter to `Attention` layer when creating `LLaMAForCausalLM` sufficient? 

[1]: https://github.com/NVIDIA/TensorRT-LLM/blob/main/examples/llama/hf_llama_convert.py
[2]: https://github.com/NVIDIA/TensorRT-LLM/blob/main/examples/llama/build.py



[1]: https://github.com/NVIDIA/TensorRT-LLM/blob/main/examples/llama/hf_llama_convert.py

**Comment by jdemouth-nvidia** [comment]
@hchoi-moveworks , there is no need to change the checkpoint conversion. The "only" thing you need to do is to propagate the `multi_block_mode` parameter. Let us know if you need help. We plan to enable `multi_block_mode` for more models in the future as that's a good possible performance optimization that we have not showcased enough, so far ;)

**Comment by tiandiao123** [comment]
what is multi-block mode? what does it actually do after setting it as multi-block?

**Comment by byshiue** [comment]
> what is multi-block mode? what does it actually do after setting it as multi-block?

The idea is same to Flash-Decoding. You could refer the paper of Flash-Decoding.

**Comment by tiandiao123** [comment]
> > what is multi-block mode? what does it actually do after setting it as multi-block?
> 
> The idea is same to Flash-Decoding. You could refer the paper of Flash-Decoding.

Thanks for clarification!

**Comment by rayleizhu** [comment]
Is this feature back-ended by the kernel here [cpp/tensorrt_llm/kernels/decoderMaskedMultiheadAttention/decoderMaskedMultiheadAttentionTemplate.h](https://github.com/NVIDIA/TensorRT-LLM/blob/release/0.5.0/cpp/tensorrt_llm/kernels/decoderMaskedMultiheadAttention/decoderMaskedMultiheadAttentionTemplate.h)?

I'm currently doing a research project on LLM inference, in which I need to customize the attention kernel slightly. The tensorrt-llm is an excellent project. It would be even better if a big picture/guideline could be provided for non-CUDA-expert researchers like me.

**Comment by byshiue** [comment]
> Is this feature back-ended by the kernel here [cpp/tensorrt_llm/kernels/decoderMaskedMultiheadAttention/decoderMaskedMultiheadAttentionTemplate.h](https://github.com/NVIDIA/TensorRT-LLM/blob/release/0.5.0/cpp/tensorrt_llm/kernels/decoderMaskedMultiheadAttention/decoderMaskedMultiheadAttentionTemplate.h)?
> 
> I'm currently doing a research project on LLM inference, in which I need to customize the attention kernel slightly. The tensorrt-llm is an excellent project. It would be even better if a big picture/guideline could be provided for non-CUDA-expert researchers like me.

If possible, we are happy to do that. But it is a little hard because the kernel is a very complicated CUDA kernel and you need many CUDA knowledge if you want to understand it and modify it. We will try our best to help answering if you encounter any question. But due to limit of resuorce, it may be delayed.

**Comment by rayleizhu** [comment]
> If possible, we are happy to do that. But it is a little hard because the kernel is a very complicated CUDA kernel and you need many CUDA knowledge if you want to understand it and modify it. We will try our best to help answering if you encounter any question. But due to limit of resuorce, it may be delayed.

1. I mean the big picture/framework of **the whole project** (instead of the details about CUDA kernel implementations), so I can know how to customize some modules.

2. Currently, I want to ensure where the backend kernels of the attention are. I get a little bit confused about the structure of the `cpp/tensorrt_llm/kernels` directory. It will be great to know how these utilities correspond to the Python front end, e.g. in the GPT and Llama model.

3. My demand requires only slight modifications on the CUDA kernels. I have a good understanding of CUDA and OpenAI Triton, FlashAttention and FlashDecoding, etc.  So, don't worry, I'm sure what I need :).

**Comment by byshiue** [comment]
Thank you for the reply. I think it is a document improvement requirement. We are continuing improveing our document, but I know it is not enough now. 

If you could provide more details about your demand, I can try providing some suggestion abuot the modules you may need to change.

**Comment by rayleizhu** [comment]
> Thank you for the reply. I think it is a document improvement requirement. We are continuing improveing our document, but I know it is not enough now.
> 
> If you could provide more details about your demand, I can try providing some suggestion abuot the modules you may need to change.

I want to return two intermediate results: the maximum logits (used to avoid numerical instability of softmax) and sum (the denominator of softmax) during softmax reduction, as in FlashAttentionV1.

**Comment by PerkzZheng** [comment]
> > Thank you for the reply. I think it is a document improvement requirement. We are continuing improveing our document, but I know it is not enough now.
> > If you could provide more details about your demand, I can try providing some suggestion abuot the modules you may need to change.
> 
> I want to return two intermediate results: the maximum logits (used to avoid numerical instability of softmax) and sum (the denominator of softmax) during softmax reduction, as in FlashAttentionV1.

@rayleizhu Hi, it should be quite easy to dump it. Let me point you to the code lines.
1. [the maximum logits (or we say max) ](https://github.com/NVIDIA/TensorRT-LLM/blob/main/cpp/tensorrt_llm/kernels/decoderMaskedMultiheadAttention/decoderMaskedMultiheadAttentionTemplate.h#L2147)
2. [the normalization factor of softmax (or we say sum)](https://github.com/NVIDIA/TensorRT-LLM/blob/main/cpp/tensorrt_llm/kernels/decoderMaskedMultiheadAttention/decoderMaskedMultiheadAttentionTemplate.h#L2185)

and each block represents one of batch x beam_width x num_heads, so it should be quite easy to let the thread0 of the block dump the max/sum values. 

You can also refer to the [multi-block implementation](https://github.com/NVIDIA/TensorRT-LLM/blob/main/cpp/tensorrt_llm/kernels/decoderMaskedMultiheadAttention/decoderMaskedMultiheadAttentionTemplate.h#L2512-L2513), which also dumps them to global memory for the final reduction. 

Let me know if you have further questions.

**Comment by byshiue** [comment]
Close this bug because the question is answered. Feel free to reopen it if needed.
